### ISSUE TITLE
[Parity 4/16] torch.distributions - all distributions, constraints, transforms, KL

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#213]
Part of epic #209.

## Summary

`torch.distributions` is a full probability DSL: ~40 distributions, constraints, bijective transforms, KL divergence registrations, and a reparameterized-sample path that is core to VAEs, policy-gradient RL, diffusion, normalizing flows, and Bayesian deep learning. We have none of it at the tensor-library level.

### Distribution classes (all with `sample`, `rsample` where applicable, `log_prob`, `entropy`, `mean`, `variance`, `stddev`, `cdf`, `icdf`)
- **Univariate continuous**: Normal, LogNormal, HalfNormal, StudentT, Cauchy, HalfCauchy, Uniform, Exponential, Laplace, Gumbel, Pareto, Weibull, Beta, Chi2, Gamma, InverseGamma, FisherSnedecor, Kumaraswamy, VonMises, ContinuousBernoulli
- **Univariate discrete**: Bernoulli, Binomial, Categorical, OneHotCategorical, Geometric, NegativeBinomial, Poisson, Multinomial
- **Multivariate**: MultivariateNormal, LowRankMultivariateNormal, Dirichlet, MixtureSameFamily, Independent, TransformedDistribution, RelaxedBernoulli, RelaxedOneHotCategorical

### Transforms (bijective, with `log_abs_det_jacobian`)
- AffineTransform, ExpTransform, SigmoidTransform, TanhTransform, SoftmaxTransform, StickBreakingTransform, LowerCholeskyTransform, CorrCholeskyTransform, PowerTransform, AbsTransform, ComposeTransform, ReshapeTransform, IndependentTransform, CumulativeDistributionTransform, CatTransform, StackTransform

### Constraints
- real, positive, nonnegative, unit_interval, simplex, positive_definite, lower_cholesky, corr_cholesky, lower_triangular, less_than, greater_than, interval, integer_interval, multinomial

### Registrations / utilities
- `kl_divergence(p, q)` with type-pair registry (100+ analytical pairs in PyTorch)
- `Independent` reinterprets batch dims as event dims
- `TransformedDistribution` = base + Transforms
- `ExpandingDistribution`, `ExponentialFamily`
- Gumbel-softmax sampling (hard + soft)

## Current state in the repo

- Nothing. No `Distributions/` directory. Random number generation lives in `Helpers/LockedRandom.cs`, `Helpers/RandomHelper.cs`, `Helpers/SimdRandom.cs`.

## How we beat PyTorch

1. **SIMD/GPU-native samplers**: PyTorch's samplers are op-by-op kernels. We implement batched samplers that vectorize across the batch dimension using SIMD (CPU) and warp-level primitives (GPU). Target 2–5× on batch sampling.
2. **Philox counter-based RNG** for deterministic, parallel-friendly sampling — matches `torch.Generator(device="cuda")` but exposed as a first-class `IPhiloxGenerator<T>` so users can hash-derive seeds per batch row (impossible in PyTorch without workarounds).
3. **Closed-form `log_prob`** via `SymbolicBackwardGraph` — we can simplify `log(exp(...) / Z)` at compile time. PyTorch runs these as dense kernels.
4. **KL divergence registry backed by `OpRegistry`**: users can register new analytic pairs per-conjugate-family. PyTorch's registry is hard-coded.
5. **Reparameterization for every eligible distribution** including implicit reparam (Figurnov 2018) for Gamma/Beta/Dirichlet — PyTorch only provides this for Normal/Student's-t/Gumbel.
6. **Sub-byte RelaxedCategorical**: sampling and `log_prob` with int4/FP4 support for quantized latent spaces (unique to us, bolts onto our quant stack).
7. **Mixed-precision sampling paths** that respect `LayerPrecisionPolicy`.
8. **Compile-mode native**: every distribution is a first-class node in `LazyTensorScope` with its own backward and fusion rules. PyTorch's distributions don't compose with `torch.compile` cleanly.
9. **Normalizing-flow composites**: ship Planar, Radial, RealNVP, MAF, NSF as `TransformedDistribution` presets — users otherwise reach for `nflows` / `zuko`.

### Framework
- [ ] `IDistribution<T>` base with `Sample`, `RSample`, `LogProb`, `Entropy`, `Mean`, `Variance`, `StdDev`, `Cdf`, `Icdf`, `Support`, `BatchShape`, `EventShape`
- [ ] `Constraint<T>` abstraction + built-in constraints above
- [ ] `Transform<T>` abstraction + all transforms above
- [ ] `ExponentialFamily<T>` helper
- [ ] `Independent<T>`, `TransformedDistribution<T>`, `MixtureSameFamily<T>`, `ExpandingDistribution<T>`
- [ ] `KLDivergence.Register<TP, TQ>(fn)` + built-in analytical pairs
- [ ] Philox generator + thread-safe integration with `SimdRandom` / `LockedRandom`

### Distributions (each: sample + rsample where applicable + log_prob + entropy + mean + variance + cdf/icdf where defined)
- [ ] Normal, LogNormal, HalfNormal, StudentT, Cauchy, HalfCauchy, Uniform, Exponential, Laplace, Gumbel, Pareto, Weibull
- [ ] Beta, Chi2, Gamma, InverseGamma, FisherSnedecor, Kumaraswamy, VonMises, ContinuousBernoulli
- [ ] Bernoulli, Binomial, Categorical, OneHotCategorical, Geometric, NegativeBinomial, Poisson, Multinomial
- [ ] MultivariateNormal (full, diagonal, low-rank), Dirichlet
- [ ] RelaxedBernoulli, RelaxedOneHotCategorical (Gumbel-softmax)

### Bonus (beats PyTorch scope)
- [ ] Planar, Radial, RealNVP, MAF, NSF flow presets
- [ ] Implicit reparam for Gamma/Beta/Dirichlet
- [ ] Sub-byte relaxed-categorical with int4/FP4 latents

## Acceptance criteria

- Property tests: `sample().mean() ≈ mean`, `.variance()` matches empirical within tolerance over N=100k samples.
- `log_prob` numerical parity vs SciPy reference on all distributions.
- Gradcheck for every differentiable `rsample` + `log_prob`.
- KL divergence: 100% of registered pairs have a test that (a) matches the analytical value and (b) matches Monte-Carlo estimate within tolerance.
- Full GPU parity on at least CUDA; other backends best-effort.
- Benchmarks: batched sampling throughput vs PyTorch; target 2× win on bulk categorical / normal / multivariate-normal sampling.

## Implementation notes

- New module `src/AiDotNet.Tensors/Distributions/`.
- Each distribution gets its own file; shared helpers in `Distributions/Helpers/`.
- Transforms and constraints in `Distributions/Transforms/` and `Distributions/Constraints/`.
- Reuse `Operators/` for `log`, `exp`, `digamma` (once `torch.special` lands in the tensor-ops issue).
- Hook `OpRegistry` so distribution ops are compile-mode friendly.

**Comment by ooples** [comment]
# Implementation plan — detailed

## 1. Current state

Empty. RNG infrastructure only: `Helpers/LockedRandom.cs`, `Helpers/RandomHelper.cs`, `Helpers/SimdRandom.cs`.

## 2. Scope — full inventory

### Framework (7 abstractions)
```
IDistribution<T> { Sample, RSample, LogProb, Entropy, Mean, Variance, StdDev, Cdf, Icdf,
                   Support, BatchShape, EventShape }
Constraint<T> + 14 built-ins (real, positive, nonnegative, unit_interval, simplex,
  positive_definite, lower_cholesky, corr_cholesky, lower_triangular, less_than,
  greater_than, interval, integer_interval, multinomial)
Transform<T> + 16 built-ins (Affine, Exp, Sigmoid, Tanh, Softmax, StickBreaking,
  LowerCholesky, CorrCholesky, Power, Abs, Compose, Reshape, Independent, CDFTransform,
  Cat, Stack)
ExponentialFamily<T> helper
Independent<T>, TransformedDistribution<T>, MixtureSameFamily<T>, ExpandingDistribution<T>
KLDivergence registry with ≥60 analytical pairs
IPhiloxGenerator<T> — counter-based deterministic RNG (CPU + GPU)
```

### Distributions (42 total)

| Category | Distributions |
|---|---|
| Univariate continuous (20) | Normal, LogNormal, HalfNormal, StudentT, Cauchy, HalfCauchy, Uniform, Exponential, Laplace, Gumbel, Pareto, Weibull, Beta, Chi2, Gamma, InverseGamma, FisherSnedecor, Kumaraswamy, VonMises, ContinuousBernoulli |
| Univariate discrete (8) | Bernoulli, Binomial, Categorical, OneHotCategorical, Geometric, NegativeBinomial, Poisson, Multinomial |
| Multivariate (4) | MultivariateNormal (full/diag/low-rank), Dirichlet, MixtureSameFamily, Independent |
| Relaxed (2) | RelaxedBernoulli, RelaxedOneHotCategorical (Gumbel-softmax soft + hard) |
| Composed | TransformedDistribution variants: Planar, Radial, RealNVP, MAF, NSF (5 flow presets — bonus) |

Each distribution ships (where defined): `Sample`, `RSample`, `LogProb`, `Entropy`, `Mean`, `Variance`, `StdDev`, `Cdf`, `Icdf`, `Support`, `BatchShape`, `EventShape`.

## 3. Per-distribution sample algorithm (representative)

| Distribution | Sample algorithm | Reparam |
|---|---|---|
| Normal | Box-Muller w/ Marsaglia polar fallback (SIMD) | z*σ + μ |
| Gamma | Marsaglia-Tsang for α≥1; scaled α+1 for α<1 | Figurnov 2018 implicit |
| Beta | via Gamma ratio | Figurnov 2018 implicit |
| Dirichlet | Gamma-ratio | Figurnov 2018 implicit |
| Categorical | Gumbel-max trick (differentiable); inverse-CDF (plain) | Gumbel-softmax |
| StudentT | Normal / √(Chi2/df) | z/√(u/df), u~Chi2 |
| VonMises | Best-Fisher rejection | No (non-reparam) |
| NegativeBinomial | Gamma-Poisson | No |
| MultivariateNormal | Cholesky + Normal mult (needs #211) | L*z + μ |

## 4. Win vectors

1. **SIMD batched samplers**: AVX-512 vectorized across batch dim for every distribution. Target 2–5× PyTorch op-by-op.
2. **GPU warp-level samplers**: each warp generates 32 samples in parallel via cuRAND/warp-shuffle. Target 2–4× bulk categorical.
3. **`IPhiloxGenerator<T>`**: counter-based RNG deterministic across CPU/GPU. Users hash-derive per-batch seeds.
4. **Symbolic-backward `log_prob`**: `AlgebraicSimplificationPass` collapses `log(exp(...) / Z)` at compile time.
5. **Extensible KL registry**: `KLDivergence.Register<TP, TQ>(fn)` — PyTorch's is hard-coded.
6. **Universal reparameterization**: implicit reparam (Figurnov 2018) for Gamma/Beta/Dirichlet. PyTorch has these gated behind "implicit_reparam" workarounds.
7. **Sub-byte RelaxedCategorical**: sampling + `log_prob` with int4/FP4 latents for quantized VAEs — unique.
8. **Mixed-precision sampling paths**: respects `LayerPrecisionPolicy`.
9. **First-class compile-mode nodes**: distributions are `LazyNode`s that fuse with surrounding ops.
10. **Flow presets**: Planar, Radial, RealNVP, MAF, NSF — users otherwise reach for `nflows`/`zuko`.

## 5. Per-axis beat targets

| Axis | Target |
|---|---|
| Bulk Normal sampling (batch=1024) | ≥2× PyTorch on CPU SIMD; ≥1.5× on CUDA |
| Bulk Categorical sampling (batch=10k, K=1000) | ≥3× PyTorch on CPU; ≥2× on CUDA |
| MultivariateNormal (d=256, batch=64) | ≥2× PyTorch (shared Cholesky + batched mult) |
| Dirichlet sampling with implicit reparam | Feature parity + gradcheck-pass (PyTorch requires workaround) |
| `log_prob` on all distributions | Parity vs SciPy within tolerance |
| KL pairs | 100% of registered match analytical + MC within tolerance |
| RelaxedCategorical int4 | Feature unique — works, no PyTorch baseline |
| All 6 GPU backends | CUDA + HIP + Metal + Vulkan + OpenCL + WebGPU samplers (managed xorshift fallback for backends without RNG libs) |
| Autograd | Gradcheck for every differentiable `rsample` + `log_prob` |
| Determinism | Same `(seed, subsequence, offset)` → identical bits CPU↔GPU |

## 6. File inventory

**New (~110):**
- `src/AiDotNet.Tensors/Distributions/` — 42 distribution files
- `src/AiDotNet.Tensors/Distributions/Transforms/` — 16 files
- `src/AiDotNet.Tensors/Distributions/Constraints/` — 14 files
- `src/AiDotNet.Tensors/Distributions/Helpers/` — IDistribution, ExponentialFamily, KLRegistry, MixtureSameFamily, Independent, TransformedDistribution (~10)
- `src/AiDotNet.Tensors/Helpers/Philox/` — IPhiloxGenerator, PhiloxCpu, PhiloxGpu bindings (per backend) (~8)
- `src/AiDotNet.Tensors/Engines/Simd/Random/` — SIMD samplers for each family (~10)
- GPU samplers per backend (~25)
- Normalizing-flow presets (~5)

**Modified (~8):** `RandomHelper.cs`, `SimdRandom.cs`, `OpRegistry.cs`, `DifferentiableOps.cs`, `LockedRandom.cs`.

**Tests (~50):** property tests, `log_prob` parity vs SciPy, gradcheck for rsample, KL pair tests (analytical + MC), device parity, determinism.

**Benchmarks (~5):** `NormalSamplingBenchmarks`, `CategoricalSamplingBenchmarks`, `MvnSamplingBenchmarks`, `LogProbBenchmarks`, `FlowsBenchmarks`.

## 7. Test matrix per distribution

| Axis | Values |
|---|---|
| parameter shapes | scalar, batched (N=), batched + event |
| dtype | fp32, fp64, fp16 (where stable) |
| sample count | 1, 100, 100k (property tests) |
| properties | empirical mean ≈ mean(); variance ≈ variance(); within tolerance |
| `log_prob` | parity vs SciPy at 100 random points |
| gradcheck | rsample + log_prob where differentiable |
| KL | analytical test + Monte Carlo test |
| device | CPU + each GPU where backend-sampler exists |
| determinism | same seed → same samples across CPU/GPU |

## 8. Acceptance gate

1. All 42 distributions pass property + log_prob + gradcheck where applicable.
2. 60+ KL pairs registered and tested both analytically + MC.
3. Philox determinism test: CPU↔CUDA identical bits.
4. Sampling throughput benchmarks meet targets.
5. All 6 GPU backends have samplers or documented skip.
6. Flow presets (Planar/Radial/RealNVP/MAF/NSF) train on a toy 2D density estimation task.
7. Sub-byte RelaxedCategorical ships with int4/FP4 support test.

## 9. Timeline — ~7 weeks

W1 Framework + Philox + Normal/Uniform/Exponential/Bernoulli/Categorical.
W2 Continuous univariate (14 more).
W3 Discrete univariate (7 more) + multivariate.
W4 Transforms + constraints + TransformedDistribution + Independent + MixtureSameFamily.
W5 KL registry + implicit reparam for Gamma/Beta/Dirichlet.
W6 Flow presets + sub-byte RelaxedCategorical + GPU samplers.
W7 Benchmarks + PR.

## 10. Dependencies

- **Blocked by**: #210 (`log`, `exp`, `digamma`, `lgamma` as Tensor ops), #211 (Cholesky for MVN), #219 (cuRAND binding for GPU Philox).
- **Unblocks**: #214 (rsample + log_prob foundations for VI, normalizing-flow training).

## 11. Risks

- **Implicit reparam gradients** (Figurnov 2018) — math is involved; must ship with ref paper link + rigorous gradcheck.
- **MVN with low-rank/PSD constraint** — backward through Cholesky at near-singular point requires ε-regularization.
- **Flow NF/NSF stability** — will need gradient-clipping defaults documented.

---

*Part of epic #209. Plan ratified 2026-04-19.*
